Basic_viewer/include/CGAL/Drawing_functor.h

```diff
@@ -0,0 +1,110 @@
+//
+// Author(s):   Guillaume Damiand <guillaume.damiand@liris.cnrs.fr>
+//              Mostafa Ashraf <mostaphaashraf1996@gmail.com>
+
+#ifndef CGAL_DRAWING_FUNCTORS_H
+#define CGAL_DRAWING_FUNCTORS_H
+
+#include <CGAL/license/GraphicsView.h>
+#include <CGAL/IO/Color.h>
+#include <functional>
+
+namespace CGAL {
+
+// Drawing functor for a 2D combinatorial data structure
+// (with vertices, edges and faces)
+template <typename DS,
+          typename vertex_handle,
+          typename edge_handle,
+          typename face_handle>
+struct Drawing_functor
+{
+  Drawing_functor(): m_enabled_vertices(true),
+                     m_enabled_edges(true),
+                     m_enabled_faces(true)
+  {
+    draw_vertex=[](const DS &, vertex_handle)->bool { return true; };
+    draw_edge=[](const DS &, edge_handle)->bool { return true; };
+    draw_face=[](const DS &, face_handle)->bool { return true; };
+
+    colored_vertex=[](const DS &, vertex_handle)->bool { return false; };
+    colored_edge=[](const DS &, edge_handle)->bool { return false; };
+    colored_face=[](const DS &, face_handle)->bool { return false; };
+
+    face_wireframe=[](const DS &, face_handle)->bool { return false; };
+  }
+
+  // The seven following functions should not be null
+  std::function<bool(const DS &, vertex_handle)> draw_vertex;
+  std::function<bool(const DS &, edge_handle)>   draw_edge;
+  std::function<bool(const DS &, face_handle)>   draw_face;
+
+  std::function<bool(const DS &, vertex_handle)> colored_vertex;
+  std::function<bool(const DS &, edge_handle)>   colored_edge;
+  std::function<bool(const DS &, face_handle)>   colored_face;
+
+  std::function<bool(const DS &, face_handle)> face_wireframe;
+
+  // These functions must be non null if the corresponding colored_XXX function
+  // returns true.
+  std::function<CGAL::IO::Color(const DS &, vertex_handle)> vertex_color;
+  std::function<CGAL::IO::Color(const DS &, edge_handle)>   edge_color;
+  std::function<CGAL::IO::Color(const DS &, face_handle)>   face_color;
+
+  void disable_vertices() { m_enabled_vertices=false; }
+  void enable_vertices() { m_enabled_vertices=true; }
+  bool are_vertices_enabled() const { return m_enabled_vertices; }
+
+  void disable_edges() { m_enabled_edges=false; }
+  void enable_edges() { m_enabled_edges=true; }
+  bool are_edges_enabled() const { return m_enabled_edges; }
+
+  void disable_faces() { m_enabled_faces=false; }
+  void enable_faces() { m_enabled_faces=true; }
+  bool are_faces_enabled() const { return m_enabled_faces; }
+
+protected:
+  bool m_enabled_vertices, m_enabled_edges, m_enabled_faces;
+};
+
+// Drawing functor for a 3D combinatorial data structure
+// (with vertices, edges, faces and volumes)
+template <typename DS,
+          typename vertex_handle,
+          typename edge_handle,
+          typename face_handle,
+          typename volume_handle>
+struct Drawing_functor_with_volume :
```

Comment:
Why not partial specializations for different number of template parameters?

Reply:
This is related to the remark of Sebastien above.
You suggest to use the same name for the two classes, with 4 or 5 template parameters?
It is ok for me.

Reply:
yes.

Reply:
ok by @sloriot

Reply:
done